src/snowflake/snowpark/_internal/packaging_utils.py

```diff
@@ -0,0 +1,376 @@
+PIP_ENVIRONMENT_VARIABLE: str = "PIP_NAME"
+IMPLICIT_ZIP_FILE_NAME: str = "zipped_packages"
+SNOWPARK_PACKAGE_NAME: str = "snowflake-snowpark-python"
+NATIVE_FILE_EXTENSIONS: Set[str] = {
+    ".pyd",
+    ".pyx",
+    ".pxd",
+    ".dll" if platform.system() == "Windows" else ".so",
+}
+
+
+def get_package_name_from_metadata(metadata_file_path: str) -> Optional[str]:
+    """
+    Loads a METADATA file from the dist-info directory of an installed Python package, finds the name and version of the
+    package. The name is found on the line containing "Name: package_name" and version can be found on the line containing
+    "Version: package_version".
+
+    Args:
+        metadata_file_path (str): The path to the METADATA file.
+
+    Returns:
+        Optional[str]: The name and (if present) version of the package formatted as f"{package}==[version]". Returns
+        None if package name cannot be found.
+    """
+
+    with open(metadata_file_path, encoding="utf-8") as metadata_file:
+        contents: AnyStr = metadata_file.read()
+        regex_results = re.search("^Name: (.*)$", contents, flags=re.MULTILINE)
+        if regex_results is None:
+            return None
+        requirement_line: str = regex_results.group(1)
+
+        regex_results = re.search("^Version: (.*)$", contents, flags=re.MULTILINE)
+        if regex_results is not None:
+            version: str = regex_results.group(1)
+            requirement_line += f"=={version}"
+
+        return requirement_line.strip().lower()
+
+
+def map_python_packages_to_files_and_folders(
+    directory: str,
+) -> Dict[Requirement, List[str]]:
+    """
+    Records correspondence between installed python packages and their folder structure, using the RECORD file present
+    in most pypi packages. We use the METADATA file to deduce the package name and version, and RECORD file to map
+    correspondence between package names and folders/files.
+
+    Args:
+        directory (str): Target folder in which pip installed the packages.
+
+    Returns:
+        Dict[Requirement, List[str]: Mapping from package to a list of unique folder/file names that correspond to it.
+    """
+
+    package_name_to_record_entries_map: Dict[Requirement, List[str]] = {}
+
+    metadata_files: List[str] = glob.glob(
+        os.path.join(directory, "*dist-info", "METADATA")
+    )
+    for metadata_file in metadata_files:
+        parent_folder: str = os.path.dirname(metadata_file)
+        package: Optional[str] = get_package_name_from_metadata(metadata_file)
+
+        if package is not None:
+            # Determine which folders or files belong to this package
+            record_file_path = os.path.join(parent_folder, "RECORD")
+            if os.path.exists(record_file_path):
+                # Get unique root folder names
+                with open(record_file_path, encoding="utf-8") as record_file:
+                    record_entries = set()
+
+                    # Read in all record entries
```

Comment:
Can you add a comment to provide an example of a line in the RECORD file?

Reply:
Sure, added